### PR TITLE
[IMP] website, web_editor: option to add columns on snippet without any

### PR DESCRIPTION
**Pull request by sdegueldre**
In the previous snippet cleanup, a lot of superfluous row > col were
removed in favour of having the bare content directly in the section,
this streamlines the interface and prevents having many useless UI
elements.

At the same time however, there are some snippet on which you might want
columns so that you can pull the content off-center and give the block a
background color independently of its containing section (namely: the
cover, title and text-block snippets)

this commit adds an option to add columns on those snippets (and remove
those columns completely if they're no longer wanted)

task-2120360

### CODE REVIEW

File: addons/website/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -1096,6 +1108,43 @@ options.registry.layout_column = options.Class.extend({
...
+    async colCount(previewMode, widgetValue, params) {
+        const count = parseInt(widgetValue);
+        let $row = this.$('> .row:not(.o_we_zero_cols_content)');
+        if (!$row.length) {
+            $row = $('<div class="row"><div class="col-lg-12"></div></div>');
+            $row.children().first().append(this.$target.children());
+            this.$target.append($row);
+        }
+        const proxy = createPropertyProxy(this, {$target: $row});
+        await proxy._updateColumnCount((count || 1) - $row.children().length);
```
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/website/static/src/js/editor/snippets.options.js]: I was initially doing it this way, but I was having to treat the zero-case specially all throughout the option. This way has the advantage that it keeps all the zero-case handling code in one place, since it is, in fact, a special case. It also involved reworking a lot of the logic of the original option because the `$target` can no longer be the row since it can't be assumed to be present. It also had the problem that I had to manually manage the UI of the 0 button of the select since it's not a top-level item in the option and can't use `_computeWidgetVisibility`.

For all those reasons I thought it would be cleaner to make a separate option that inherits from the original, at the (IMO lesser) cost of duplicating a bit of XML. I had initially done as you're suggesting and I found it very messy and that it added a lot of complexity to the code, but to be fair I was very tired that day and I would probably be able to get it a lot cleaner now that my brain is back to normal. The only remaining problem would be manual UI management for the 0 item in the select, which may be preferable to the XML duplication that I have in this version.

As far as using a proxy, it was is in the same spirit as the `data-apply-to` attribute: since the method is async, I didn't want to have the same problem as we have with `_super` where the `$target` would be incorrect in asynchronous code.

What I didn't realize at the time is that since option methods are executed in a mutex, and they cannot run in parallel, we probably didn't need a proxy in the first place and we could just patch `$target` before the call and restore it after (and in this case we could do the same). The only exception would be if someone had code that ran *after* the option method returned, for example in a `setTimeout` or any other callback-based function that is not awaited within the option method, but that seems like it's not something we want to allow since the whole point of awaiting option methods is to ensure they finished their work.

So here's what I would like to do if that's ok with you:
* 1 commit for making `onCloneSnippet` and `onRemoveSnippet` take a `onSuccess` or `callback` param so that they can be awaited
* 1 commit to remove the use of property proxies since they're basically there to support a use case that we explicitly don't support (multiple option methods running in parallel)
* 1 commit to add the zero column capability to the existing layout_column option

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: - Commit 1: yes :+1: 
- Commit 2: Are you sure it was the only reason for using a proxy? Because when you did it, I did suggest to only change `$target` for the duration of the option method call and I don't think you convinced me just with the option-in-parallel argument... :/ but maybe I forgot, the mutex did appear for 13.2 not 13.1 so it is possible. (Not: in fact, if `apply-to` matches multiple elements, the options could still run in parallel for each element, justifying the use of the proxy but maybe it is a bad reason).
- Commit 3: Yes, I don't think it will be messy to go with an unique option Class handling both cases. The `$target` becomes the `.container` and it is just about finding/creating the row and leaving the old code work on that row. About the XML duplication, it is not a problem, you can leave it as a duplication for two different option instances, using the same option Class. So:
```xml
<div data-js="layout_column" data-selector="XXX" data-target=".container:not(.somethingSpecific)">
    <we-select>... choices without 0 available ...</we-select>
</div>
<div data-js="layout_column" data-selector="XXX" data-target=".container.somethingSpecific" ...>
    <we-select>... choices with 0 available ...</we-select>
</div>
```
and you can just avoid duplication of the XML with some t-set/t-raw/t-call magic. But we do need a way to have a `computeVisibility` of internal widgets of a select, but this can be a workaround for now.

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 2]
```diff
@@ -2098,8 +2101,11 @@ var SnippetsMenu = Widget.extend({
      */
     _onRemoveSnippet: function (ev) {
         ev.stopPropagation();
-        this._createSnippetEditor(ev.data.$snippet).then(function (editor) {
-            editor.removeSnippet();
+        this._createSnippetEditor(ev.data.$snippet).then(async editor => {
+            await editor.removeSnippet();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js]: @sdegueldre Note: in fact that method is not yet async but will be if I merge this: https://github.com/odoo/odoo/pull/43918 ?

Comment (sdegueldre) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: There's been a lot of code shuffling between this and the column number fix so I must've lost that somewhere. Same can be said for the `_onCloneSnippet` method, I made both of those async at some point but the code got lost somewhere. I can make both of those async and take the `onSuccess` param on the event in the first commit we discussed. As a side note, I've found myself adding this mechanism to events quite a few times, and I've seen it done by existing code quite a bit too. Might want to consider baking it into the event system at some point?

File: addons/website/static/src/js/editor/snippets.options.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/website/static/src/js/editor/snippets.options.js]: > Note: in fact, if apply-to matches multiple elements, the options could still run in parallel for each element, justifying the use of the proxy but maybe it is a bad reason

They could (and at the moment, they do), but considering the initial idea was to mimic as closely as possible the behaviour of an option method on a different target, it seems like a bad idea in retrospect. Also, looking at the code right now there's a nasty bug waiting to happen:

```javascript
if (params.applyTo) {
    const $subTargets = this.$(params.applyTo);
    const proms = _.each($subTargets, subTargetEl => {
        const proxy = createPropertyProxy(this, {$target: $(subTargetEl)});
        return this[methodName].call(proxy, previewMode, widgetValue, params);
    });
    await Promise.all(proms);
}
```

It's using `each` instead of `map`, so the promises are never awaited.

I'm pretty sure my initial intention was to avoid replicating the `this._super` behaviour where you have to cache it before any async work and I just didn't take into account that the behaviour was substantially different. If you can think of or recall any other reason why we might want to use a proxy, I'm perfectly fine with leaving it in, I'd just rather not leave it there if it's not needed as it makes it adds a level of indirection when reasoning about options.

File: addons/web_editor/static/src/js/editor/snippets.editor.js (hunk 2 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: Not sure what you would bake. Mutex-alize event handlers calls automatically per widget? Dangerous because you will not be able to preventDefault / stopPropagation correctly/intuitively. If the async part of an event handler must be waited, it seems better to make it explicit for now. And for some cases the 'async' word in the event declaration is enough to make sure a specific handler is "mutex-alized" with himself.

In any case, no point in trying to improve the old framework. The next step at improving widgets is probably using owl on our side (at least in the editor).

File: addons/website/static/src/js/editor/snippets.options.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: Ok, then let's remove the proxies in master. And you can make a PR to fix that each/map fail in the oldest stable version (probably 13.2 ?), thanks :)

[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/website/static/src/js/editor/snippets.options.js]: each/map: #44959

File: addons/website/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -1028,8 +1039,19 @@ options.registry.layout_column = options.Class.extend({
      * @see this.selectClass for parameters
      */
     selectCount: async function (previewMode, widgetValue, params) {
+        let $row = this.$('> .row:not(.o_we_zero_cols_content)');
+        if (!$row.length) {
+            this.$target.contents().wrapAll($('<div class="row"><div class="col-lg-12"/></div>'));
+            $row = this.$('> .row');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: Instead of wrapping in a row then searching the row, can't you create the row and wrap in it, avoid the search?
```javascript
$row = $('<div class="row"><div class="col-lg-12"/></div>');
this.$target.contents().wrapAll($row);
```
Mmh... maybe strange to read `wrapAll($row)`... but jquery is strange

File: addons/website/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -1028,8 +1039,19 @@ options.registry.layout_column = options.Class.extend({
...
+        }
+
         const nbColumns = parseInt(widgetValue);
-        await this._updateColumnCount(nbColumns - this.$target.children().length);
+        await this._updateColumnCount($row, (nbColumns || 1) - $row.children().length);
+        if (nbColumns === 0) {
+            $row.contents().unwrap()
+                .contents().unwrap()
+                .addClass('o_we_zero_cols_content');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: Can't you reverse the logic and have a class on row to consider instead of adding a class on every content a container without row can have?

File: addons/website/static/src/js/editor/snippets.options.js [hunk 5]
```diff
@@ -1050,41 +1072,42 @@ options.registry.layout_column = options.Class.extend({
...
      */
-    _updateColumnCount: async function (count) {
+    _updateColumnCount: async function ($row, count) {
         if (!count) {
             return;
         }
-
-        this.trigger_up('request_history_undo_record', {$target: this.$target});
+        // TODO: check that this is in fact needed
+        this.trigger_up('request_history_undo_record', {$target: $row});
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: Yep, don't understand why that would be needed either, this is part of the option system.

File: addons/website/views/snippets/snippets.xml [hunk 6]
```diff
@@ -219,8 +219,9 @@
 
     <div data-js="layout_column"
         data-selector="section"
-        data-target="> * > .row:not(.s_nb_column_fixed)">
+        data-target="> *:has(.row:not(.s_nb_column_fixed)), > .o_we_zero_cols">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml]: - Are you not missing a `>` before `.row`, like before ?
- I find the name o_we_zero_cols confusing (since there are containers with 0 cols that do not have that class), could not it be something like allow_columns or something like that?

File: addons/website/static/src/js/editor/snippets.options.js (hunk 4 above)
Comment (sdegueldre) [reply addons/website/static/src/js/editor/snippets.options.js]: This is for the hypothetical case where we allow unwrapping a row-in-col snippet (like the features grid snippet), but since we don't allow that at the moment it might not be needed at all. And if we do need it, we can indeed add a class like `o_we_no_unwrap` on the rows of the snippet in question.

File: addons/website/static/src/js/editor/snippets.options.js (hunk 5 above)
Comment (sdegueldre) [reply addons/website/static/src/js/editor/snippets.options.js]: Seems like that line was introduced with the layout_column option when it was first created. And works fine without it, I'm removing it then.

File: addons/website/views/snippets/snippets.xml (hunk 6 above)
Comment (sdegueldre) [reply addons/website/views/snippets/snippets.xml]: Considering that at the moment it's only used to allow adding column where there are none, that seems like a better name. The spec was initially framed as allowing to remove columns where they already existed so that name made more sense then, I'll change it :)

File: addons/website/static/src/js/editor/snippets.options.js (hunk 3 above)
Comment (sdegueldre) [reply addons/website/static/src/js/editor/snippets.options.js]: `wrapAll` makes a copy of the wrapping element, so in your code `$row` is never actually inserted into the DOM. I could do `$row = this.$target.contents().wrapAll($('<div class="row"><div class="col-lg-12"/></div>')).parent();` however, since `wrapAll` returns the wrapped elements. Not sure that's any better though.

Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: Keep your implementation then :+1:

File: addons/website/views/snippets/snippets.xml [hunk 7]
```diff
@@ -219,8 +219,9 @@
 
     <div data-js="layout_column"
         data-selector="section"
-        data-target="> * > .row:not(.s_nb_column_fixed)">
+        data-target="> *:has(> .row:not(.s_nb_column_fixed)), > .o_we_allow_columns">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml]: In fact, the class should not contain `we` since it does not come from web_editor. So either `o_allow_columns` or `s_allow_columns` since it is something specific to snippets (like the `s_nb_column_fixed` here) :+1:

File: addons/website/static/src/js/editor/snippets.options.js [hunk 8]
```diff
@@ -1028,8 +1039,16 @@ options.registry.layout_column = options.Class.extend({
      * @see this.selectClass for parameters
      */
     selectCount: async function (previewMode, widgetValue, params) {
+        let $row = this.$('> .row');
+        if (!$row.length) {
+            $row = this.$target.contents().wrapAll($('<div class="row"><div class="col-lg-12"/></div>')).parent().parent();
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: I preferred the old way but don't really care :)

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 9]
```diff
@@ -137,7 +137,9 @@ var SnippetEditor = Widget.extend({
      * @override
      */
     destroy: function () {
+        const $optionsSection = this.$el.data('$optionsSection');
         this._super(...arguments);
+        $optionsSection.remove();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js]: Probably the same bug in 13.0, so you can make a dedicated commit in another PR. Also instead of using `.data('uglyStuff')`, you can remove the last element in `this._customize$Elements`.

Comment (sdegueldre) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: Doesn't `_customize$Elements` also contain the parent's `$optionsSection`?

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: `_customize$Elements` is a list containing the parent ones and the current one. The last one is the current one.

Comment (sdegueldre) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: Makes sense, will make a PR for that in 13.0 and link it here.

**Comment by sdegueldre** [comment]
@qsm-odoo alright so apparently everything works as intended, and I just didn't understand the js execution order as much as I thought.

using `await Promise.resolve()` puts the rest of the execution at the back of the microtask queue, which needs to be empty for the next task in the task queue to be processed, `await new Promise(resolve => setTimeout(resolve))` puts the rest of the execution at the back of the task queue. Each handler for an event is considered a different task in the queue, and so in this case, if I want to wait for the event handlers on the parent to fire, I actually need to go to the back of the task queue.

For this I used the timeout method described above, it looks like a hack but it's actually not. I put an explanation in a comment above the setTimeout, let me know if you can think of a better way to achieve this.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 10]
```diff
@@ -1028,8 +1039,23 @@ options.registry.layout_column = options.Class.extend({
...
+        }
+
         const nbColumns = parseInt(widgetValue);
-        await this._updateColumnCount(nbColumns - this.$target.children().length);
+        await this._updateColumnCount($row, (nbColumns || 1) - $row.children().length);
+        // Yield UI thread to wait for event to bubble before activate_snippet is called.
+        // In this case this lets the select handle the click event before we switch snippet.
+        await new Promise(resolve => setTimeout(resolve));
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: @sdegueldre Seems legit following your explanation. Maybe this could be done more generically though? In the activate_snippet handler at least, before option method call maybe?

Comment (sdegueldre) [reply addons/website/static/src/js/editor/snippets.options.js]: Alright, pushed the following changes:

* Using the above code in `_onActivateSnippet` handler with a more detailed comment
* Changed `_onDeactivateSnippet` to call `_onActivateSnippet` instead of calling `_activateSnippet` directly
* Changed the `removeSnippet` method to take an additional parameter. Historically, it would only be called by the left panel trash can, and so it would always need to activate the parent of the removed snippet. Since the column option now also fires a `remove_snippet` event to remove some of its contents, we don't always want to activate the parent snippet, so you can now call it with a parameter to tell it to keep the current active snippet. This was causing other issues.

also left a few TODO comments for things to check with you.

Comment (sdegueldre) [reply addons/website/static/src/js/editor/snippets.options.js]: @qsm-odoo actually this needs some more work. I've ran into a few issues that are still not sorted out. In `removeSnippet`, the editor calls its destroy method, but the snippetsMenu keeps a reference to it in its `snippetEditors` field, and keeps trying to toggle its options after its been destroyed. At the moment, this is silent but as soon as snippetEditors start removing their `$optionsSection` properly, this will crash because `_activateSnippet` will try to call `toggleOptions(false)` on all editors (even destroyed ones), which tries to get all editors on its `customize$Elements` through jQuery's `data` method, but because the editor destroyed its `$optionsSection` (which is a `customize$Element`) it will throw. Asking the `SnippetsMenu` to destroy the editor instead (allowing it to remove it from its `snippetEditors` field) prevents the crash, but I still need to solve the "also delete all parents that are now empty" problem.

Comment (sdegueldre) [reply addons/website/static/src/js/editor/snippets.options.js]: @qsm-odoo as discussed in person, since making things more generic is more involved than we previously thought, I'll just go with the initial version of using setTimeout directly in the option method, and we'll improve the underlying system separately. If there is nothing else this PR is ready :+1:

**Comment by qsm-odoo** [comment]
Tested, result is nice :)
@robodoo r+
